Objects/memoryobject.c

```diff
@@ -3351,53 +3398,74 @@ memoryiter_next(PyObject *self)
 static PyObject *
 memory_iter(PyObject *seq)
 {
-    if (!PyMemoryView_Check(seq)) {
-        PyErr_BadInternalCall();
-        return NULL;
-    }
-    PyMemoryViewObject *obj = (PyMemoryViewObject *)seq;
-    int ndims = obj->view.ndim;
-    if (ndims == 0) {
-        PyErr_SetString(PyExc_TypeError, "invalid indexing of 0-dim memory");
+    return memoryiter_new(seq, 0);
+}
+
+PyTypeObject _PyMemoryIter_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "memory_iterator",
+    .tp_basicsize = sizeof(memoryiterobject),
+    .tp_dealloc = memoryiter_dealloc,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_traverse = memoryiter_traverse,
+    .tp_iter = PyObject_SelfIter,
+    .tp_iternext = memoryiter_next,
+};
+
+/**************************************************************************/
+/*                      Memoryview Reversed Iterator                      */
+/**************************************************************************/
+
+static PyObject *
+memorview_reverse_iternext(PyObject *self)
+{
+    memoryiterobject *it = _PyMemoryViewIter_CAST(self);
+    PyMemoryViewObject *seq = it->it_seq;
+    if (seq == NULL) {
         return NULL;
     }
-    if (ndims != 1) {
-        PyErr_SetString(PyExc_NotImplementedError,
-            "multi-dimensional sub-views are not implemented");
-        return NULL;
+    if (it->it_index >= 0 && it->it_index < it->it_length) {
+        // FEAT(picnixz): can we omit "it->it_index < it->it_length"?
+        return memoryiter_iter_nth(seq, it->it_index--, it->it_fmt);
     }
+    it->it_seq = NULL;
+    Py_DECREF(seq);
```

Comment:
Use `Py_CLEAR()`: clearing references is delicate if the object doesn't expect it to be `NULL`.
```suggestion
    Py_CLEAR(it->it_seq);
```

Reply:
Not sure if we can have arbitrary code being called here but I'll do it anyway.